lib/puppetserver/ca/action/list.rb

```diff
@@ -78,42 +87,67 @@ def run(input)
           all_certs = get_all_certs(puppet.settings).select { |cert| filter_names.call(cert) }
           requested, signed, revoked = separate_certs(all_certs)
           missing = certnames - all_certs.map { |cert| cert['name'] }
+          
+          if all && output_format == 'json'
+            grouped_cert = { "requested" => requested,
+                             "signed" => signed, 
+                             "revoked" => revoked,
+                             "missing" => missing }.to_json
+            @logger.inform(grouped_cert)
+            return 0 
+          end
```

Comment:
Do folks have thoughts on whether it makes sense to always have all the keys in the JSON output, even if some of the lists are empty? That maybe feels like better JSON etiquette, but doesn't match what we're doing for text output.

If we don't want keys with empty lists, then I think it might make sense to incorporate this into `output_certs_by_format`, and only add keys to the has that actually have certs associated.

Reply:
Yeah I can make that change.  Am I waiting on feedback from folks on how they want to do this?

Reply:
I don't have a strong opinion on the format (empty array values vs no keys where there'd be empty array values), however I do think that we should refactor the printing to take place in one spot so that shared behaviors are near each other and we aren't returning from the middle of a large method.

Reply:
Agreed. Whatever's easiest then, and we should make sure to test that it does the right thing (either skip keys or put empty arrays in them).

Reply:
I made changes so now that empty lists don't show up with the `--all` flag. Also, delegated the large chunks of logic down to the `output_certs_by_state`.